ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -2,60 +2,290 @@ import { drawOccupancyZonesTexts } from './drawOccupancyZonesTexts';
+  ctx.fill();
+  ctx.stroke();
+  ctx.fillStyle = WHITE_100;
+  ctx.font = SANS;
+  ctx.textAlign = 'center';
+  ctx.textBaseline = 'middle';
+  ctx.fillText(
+    `+${remainingTrainsNb} trains`,
+    xPosition + REMAINING_TRAINS_WIDTH / 2,
+    textY + REMAINING_TRAINS_HEIGHT / 2
+  );
+};
+
+const drawOccupationZone = ({
+  ctx,
+  zone,
+  arrivalTimePixel,
+  departureTimePixel,
+  yPosition,
+  isThroughTrain,
+  selectedTrainId,
+}: {
+  ctx: CanvasRenderingContext2D;
+  zone: OccupancyZone;
+  arrivalTimePixel: number;
+  departureTimePixel: number;
+  yPosition: number;
+  isThroughTrain: boolean;
+  selectedTrainId: string;
+}) => {
+  ctx.fillStyle = zone.color;
+  ctx.strokeStyle = WHITE_100;
+  ctx.lineJoin = 'round';
+  ctx.lineCap = 'round';
+  ctx.font = '400 10px IBM Plex Mono';
+
+  const extraWidth = isThroughTrain ? X_TROUGHTRAIN_BACKGROUND_PADDING : X_BACKGROUND_PADDING;
+  const { originTextLength, destinationTextLength } = {
+    originTextLength: ctx.measureText(zone.originStation || '--').width,
+    destinationTextLength: ctx.measureText(zone.destinationStation || '--').width,
+  };
+
+  if (selectedTrainId === zone.id) {
+    ctx.save();
+    ctx.fillStyle = SELECTION_20;
+    ctx.beginPath();
+    ctx.roundRect(
+      arrivalTimePixel - originTextLength - extraWidth,
+      yPosition - BACKGROUND_HEIGHT / 2,
+      departureTimePixel -
+        arrivalTimePixel +
+        originTextLength +
+        destinationTextLength +
+        extraWidth * 2,
+      BACKGROUND_HEIGHT,
+      SELECTED_TRAIN_ID_GRADIANT
+    );
+    ctx.fill();
+    ctx.restore();
+  }
+
+  if (isThroughTrain) {
+    drawThroughTrain({ ctx, arrivalTimePixel });
+  } else {
+    drawDefaultZone({ ctx, arrivalTimePixel, departureTimePixel, yPosition });
+  }
+
+  drawOccupancyZonesTexts({
+    ctx,
+    zone,
+    arrivalTimePixel,
+    departureTimePixel,
+    yPosition,
+    isThroughTrain,
+    selectedTrainId,
+  });
+};
+
 export const drawOccupancyZones = ({
   ctx,
   width,
   height,
   tracks,
   occupancyZones,
   getTimePixel,
+  selectedTrainId,
 }: {
   ctx: CanvasRenderingContext2D;
   width: number;
   height: number;
   tracks: Track[] | undefined;
   occupancyZones: OccupancyZone[] | undefined;
   getTimePixel: (time: number) => number;
+  selectedTrainId: string;
 }) => {
   ctx.clearRect(0, 0, width, height);
 
-  tracks?.forEach((track, index) => {
+  if (!tracks || !occupancyZones || occupancyZones.length === 0) return;
+
+  const sortedOccupancyZones = occupancyZones.sort(
+    (a, b) => a.arrivalTime.getTime() - b.arrivalTime.getTime()
+  );
+
+  tracks.forEach((track, index) => {
     const trackTranslate = index === 0 ? CANVAS_PADDING : TRACK_HEIGHT_CONTAINER;
     ctx.translate(0, trackTranslate);
 
-    const trackOccupancyZones = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const filteredOccupancyZones = sortedOccupancyZones.filter((zone) => zone.trackId === track.id);
```

Comment:
I think you can add this early return. I think the code might break if there is no zone for a track
```suggestion
    const filteredOccupancyZones = sortedOccupancyZones.filter((zone) => zone.trackId === track.id);
    
    if (!filteredOccupancyZones) continue;
```

Reply:
done at start of the while loop